### PR TITLE
Replace NULLs with nullptr

### PR DESCRIPTION
**Pull request by egecetin**
- Replaced all NULL occurences with find and replace
- Fixed typo in a protected function inside DnsLayer getNexResource -> getNextResource
- Fixed minor clang-tidy warnings (2 unused variable set removed)

### CODE REVIEW

File: Examples/DpdkExample-FilterTraffic/main.cpp [hunk 1]
```diff
@@ -44,22 +44,22 @@
...
+	{ "match-dest-port",      optional_argument, nullptr, 'P' },
+	{ "match-protocol",       optional_argument, nullptr, 'o' },
+	{ "core-mask",            optional_argument, nullptr, 'c' },
+	{ "mbuf-pool-size",       optional_argument, nullptr, 'm' },
+	{ "rx-queues",            optional_argument, nullptr, 'r' },
+	{ "tx-queues",            optional_argument, nullptr, 't' },
+	{ "help",                 optional_argument, nullptr, 'h' },
+	{ "version",              optional_argument, nullptr, 'v' },
+	{ "list",                 optional_argument, nullptr, 'l' },
+	{ nullptr,	              0,                 nullptr,  0  }
```
Comment (egecetin) [inline Examples/DpdkExample-FilterTraffic/main.cpp:62]: Fix this

File: Packet++/header/SSLCommon.h [hunk 2]
```diff
@@ -323,7 +323,7 @@ namespace pcpp
 	 */
 	enum SSLKeyExchangeAlgorithm
 	{
-		/** NULL value */
+		/** nullptr value */
```
Comment (egecetin) [inline Packet++/header/SSLCommon.h:326]: Fix this

Comment (egecetin) [reply Packet++/header/SSLCommon.h]: Replaced "NULL" with "Null" to do not pollute future "NULL" searches

**Review by tigercosmos — APPROVED** [review]
LGTM

File: Packet++/header/DnsLayerEnums.h [hunk 3]
```diff
@@ -31,7 +31,7 @@ namespace pcpp
 		DNS_TYPE_MG,
 		/** mail rename domain name record */
 		DNS_TYPE_MR,
-		/** NULL record */
+		/** nullptr record */
```
Comment (egecetin) [inline Packet++/header/DnsLayerEnums.h:34]: Fix me

File: Examples/IcmpFileTransfer/IcmpFileTransfer-pitcher.cpp [hunk 4]
```diff
@@ -36,8 +36,8 @@ void usleep(__int64 usec)
 
 	ft.QuadPart = -(10 * usec);  // Convert to 100 nanosecond interval, negative value indicates relative time
 
-	timer = CreateWaitableTimer(NULL, TRUE, NULL);
-	SetWaitableTimer(timer, &ft, 0, NULL, NULL, 0);
+	timer = CreateWaitableTimer(nullptr, TRUE, nullptr);
+	SetWaitableTimer(timer, &ft, 0, nullptr, nullptr, 0);
```
Comment (seladb) [inline Examples/IcmpFileTransfer/IcmpFileTransfer-pitcher.cpp:40]: I don't know if it's a good idea to replace `NULL` with `nullptr` when calling Windows API. People have different opinions on this:
https://www.reddit.com/r/cpp/comments/82p4bx/when_msdn_says_null_is_it_okay_to_use_nullptr/?rdt=56558
https://devblogs.microsoft.com/oldnewthing/20180307-00/?p=98175

Comment (tigercosmos) [reply Examples/IcmpFileTransfer/IcmpFileTransfer-pitcher.cpp:40]: But keeping `NULL` is also annoying, and the CI seems fine.

Comment (seladb) [reply Examples/IcmpFileTransfer/IcmpFileTransfer-pitcher.cpp:40]: The CI doesn't test `IcmpFileTransfer` example because it's not easy to test it in CI.
The build passes, but it doesn't mean the Windows API calls work the same with `NULL` / `0` and `nullptr`

File: Packet++/header/DnsLayerEnums.h (hunk 3 above)
Comment (seladb) [reply Packet++/header/DnsLayerEnums.h:34]: Yes, I think we should keep this one `NULL`

File: Packet++/header/NullLoopbackLayer.h [hunk 5]
```diff
@@ -26,7 +26,7 @@ namespace pcpp
 
 	/**
 	 * @class NullLoopbackLayer
-	 * Represents a NULL/Loopback layer
+	 * Represents a nullptr/Loopback layer
```
Comment (seladb) [inline Packet++/header/NullLoopbackLayer.h:29]: This should be NULL

File: Packet++/header/PPPoELayer.h [hunk 6]
```diff
@@ -361,23 +362,23 @@ namespace pcpp
 		/**
 		 * Get a PPPoE Tag by tag type.
 		 * @param[in] tagType The type of the tag to search
-		 * @return A PPPoETag object that contains the first tag that matches this type, or logical NULL
+		 * @return A PPPoETag object that contains the first tag that matches this type, or logical nullptr
```
Comment (seladb) [inline Packet++/header/PPPoELayer.h:365]: I would revert to `logical null` rather than `logical nullptr`. Ditto in the rest of the changes in this file

File: Packet++/header/RadiusLayer.h [hunk 7]
```diff
@@ -249,24 +249,24 @@ namespace pcpp
 
 		/**
 		 * @return The first RADIUS attribute in the packet. If there are no attributes the returned value will contain
-		 * a logical NULL (RadiusAttribute#isNull() == true)
+		 * a logical nullptr (RadiusAttribute#isNull() == true)
```
Comment (seladb) [inline Packet++/header/RadiusLayer.h:252]: ditto: revert to `logical null` here and other places in this file

File: Packet++/header/TLVData.h [hunk 8]
```diff
@@ -305,12 +305,12 @@ namespace pcpp
 		 * @param[in] tlvDataBasePtr A pointer to the TLV data byte stream
 		 * @param[in] tlvDataLen The TLV data byte stream length
 		 * @return An instance of type TLVRecordType that wraps the record following the record given as input. If the
-		 * input record.isNull() is true or if the next record is out of bounds of the byte stream, a logical NULL
+		 * input record.isNull() is true or if the next record is out of bounds of the byte stream, a logical nullptr
```
Comment (seladb) [inline Packet++/header/TLVData.h:308]: ditto: replace `logical nullptr` to `logical null` here and in other changes in this file

File: Packet++/header/TcpLayer.h [hunk 9]
```diff
@@ -549,23 +549,23 @@ namespace pcpp
 		/**
 		 * Get a TCP option by type
 		 * @param[in] option TCP option type to retrieve
-		 * @return An TcpOption object that contains the first option that matches this type, or logical NULL
+		 * @return An TcpOption object that contains the first option that matches this type, or logical nullptr
```
Comment (seladb) [inline Packet++/header/TcpLayer.h:552]: ditto

File: Packet++/header/DhcpLayer.h [hunk 10]
```diff
@@ -765,23 +765,23 @@ namespace pcpp
 
 		/**
 		 * @return The first DHCP option in the packet. If there are no DHCP options the returned value will contain
-		 * a logical NULL (DhcpOption#isNull() == true)
+		 * a logical nullptr (DhcpOption#isNull() == true)
```
Comment (seladb) [inline Packet++/header/DhcpLayer.h:768]: ditto

File: Packet++/header/DhcpV6Layer.h [hunk 11]
```diff
@@ -330,24 +330,24 @@ namespace pcpp
 
 		/**
 		 * @return The first DHCPv6 option in the packet. If there are no DHCPv6 options the returned value will contain
-		 * a logical NULL (DhcpV6Option#isNull() == true)
+		 * a logical nullptr (DhcpV6Option#isNull() == true)
```
Comment (seladb) [inline Packet++/header/DhcpV6Layer.h:333]: ditto

File: Packet++/header/IPv4Layer.h [hunk 12]
```diff
@@ -594,23 +594,24 @@ namespace pcpp
 		/**
 		 * Get an IPv4 option by type.
 		 * @param[in] option IPv4 option type
-		 * @return An IPv4Option object that contains the first option that matches this type, or logical NULL
+		 * @return An IPv4Option object that contains the first option that matches this type, or logical nullptr
```
Comment (seladb) [inline Packet++/header/IPv4Layer.h:597]: ditto

File: Packet++/header/IPv6Extensions.h [hunk 13]
```diff
@@ -333,23 +333,23 @@ namespace pcpp
 		/**
 		 * Retrieve an option by its type
 		 * @param[in] optionType Option type
-		 * @return An IPv6Option object that wraps the option data. If option isn't found a logical NULL is returned
+		 * @return An IPv6Option object that wraps the option data. If option isn't found a logical nullptr is returned
```
Comment (seladb) [inline Packet++/header/IPv6Extensions.h:336]: ditto

File: Packet++/header/NdpLayer.h [hunk 14]
```diff
@@ -131,30 +131,30 @@ namespace pcpp
 		/**
 		 * Get a NDP option by type.
 		 * @param[in] option NDP option type
-		 * @return An NdpOption object that contains the first option that matches this type, or logical NULL
+		 * @return An NdpOption object that contains the first option that matches this type, or logical nullptr
```
Comment (seladb) [inline Packet++/header/NdpLayer.h:134]: ditto

File: Packet++/src/DnsResource.cpp [hunk 15]
```diff
@@ -180,7 +180,6 @@ namespace pcpp
 				// set the pointer to the encoded string result
 				result[0] = (uint8_t)0xc0;
 				result[1] = (uint8_t)pointerInPacket;
-				result += 2;
```
Comment (seladb) [inline Packet++/src/DnsResource.cpp:183]: Why was this change added?

File: Packet++/src/TcpReassembly.cpp [hunk 16]
```diff
@@ -515,8 +515,6 @@ namespace pcpp
 			PCPP_LOG_DEBUG(
 			    "Starting first iteration of checkOutOfOrderFragments - looking for fragments that match the current sequence or have smaller sequence");
 
-			foundSomething = false;
```
Comment (seladb) [inline Packet++/src/TcpReassembly.cpp:518]: Why was this change added?

File: Packet++/src/TextBasedProtocol.cpp [hunk 17]
```diff
@@ -306,7 +306,7 @@ namespace pcpp
 		// update offsets of all fields after this field
 		HeaderField* curField = fieldToRemove->getNextField();
 		shiftFieldsOffset(curField, 0 - fieldToRemove->getFieldSize());
-		//	while (curField != NULL)
+		//	while (curField != nullptr)
```
Comment (seladb) [inline Packet++/src/TextBasedProtocol.cpp:309]: This is a part of comment, we can probably remove all the commented-out code

File: Pcap++/header/PcapLiveDevice.h [hunk 18]
```diff
@@ -274,7 +274,7 @@ namespace pcpp
 		}
 
 		/**
-		 * @return A human-readable description of the device, taken from pcap_if_t->description. May be NULL in some
+		 * @return A human-readable description of the device, taken from pcap_if_t->description. May be nullptr in some
```
Comment (seladb) [inline Pcap++/header/PcapLiveDevice.h:277]: This is a string so cannot be `nullptr`. We should probably replace with `May be an empty string in some interfaces`

File: Pcap++/src/RawSocketDevice.cpp [hunk 19]
```diff
@@ -430,7 +430,7 @@ namespace pcpp
 
 		int n = 1;
 		DWORD dwBytesRet;
-		if (WSAIoctl(fd, SIO_RCVALL, &n, sizeof(n), NULL, 0, &dwBytesRet, NULL, NULL) == SOCKET_ERROR)
+		if (WSAIoctl(fd, SIO_RCVALL, &n, sizeof(n), nullptr, 0, &dwBytesRet, nullptr, nullptr) == SOCKET_ERROR)
```
Comment (seladb) [inline Pcap++/src/RawSocketDevice.cpp:433]: same comment regarding Windows API

File: Examples/IcmpFileTransfer/IcmpFileTransfer-pitcher.cpp (hunk 4 above)
Comment (Dimi1010) [reply Examples/IcmpFileTransfer/IcmpFileTransfer-pitcher.cpp:40]: Generally, `nullptr` tends to work fine in most cases, but I do recommend testing it for correct behavior. 

Personally, I use the `NULL` macro when it is in the Win32 API call directly as the calls already have macros sprinkled in like `INFINITE` so it feels more consistent to me and use `nullptr` in all other cases, like assigning to an `LPxxx` variable. ¯\\\_(ツ)\_/¯

Comment (tigercosmos) [reply Examples/IcmpFileTransfer/IcmpFileTransfer-pitcher.cpp:40]: Btw, maybe we should have a list of untested code... Also better to have a CI guard that warns when we touch the area without test...

Comment (seladb) [reply Examples/IcmpFileTransfer/IcmpFileTransfer-pitcher.cpp:40]: @tigercosmos you can see in codecov what's not covered: https://app.codecov.io/gh/seladb/PcapPlusPlus

Comment (tigercosmos) [reply Examples/IcmpFileTransfer/IcmpFileTransfer-pitcher.cpp:40]: It's not straightforward though...

File: Packet++/src/DnsResource.cpp (hunk 15 above)
Comment (egecetin) [reply Packet++/src/DnsResource.cpp:183]: It was a clang-tidy warning. This addition is unused

File: Packet++/src/TcpReassembly.cpp (hunk 16 above)
Comment (egecetin) [reply Packet++/src/TcpReassembly.cpp:518]: Same. A clang-tidy warning for unused

**Comment by egecetin** [comment]
Fixed all comment if not missed anything

File: Examples/IcmpFileTransfer/IcmpFileTransfer-pitcher.cpp (hunk 4 above)
Comment (seladb) [reply Examples/IcmpFileTransfer/IcmpFileTransfer-pitcher.cpp]: @tigercosmos what do you suggest? Not sure it's easy to maintain a list of untested code, that's why we have codecov for... 🤔

**Review by tigercosmos — APPROVED** [review]

**Review by seladb — APPROVED** [review]
